assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js

```diff
@@ -69,80 +95,57 @@ const usePaymentMethodRegistration = (
 	useEffect( () => {
-		// if all payment methods are initialized then bail.
-		if ( isInitialized ) {
-			return;
-		}
-		const updatePaymentMethod = ( current, canPay = true ) => {
-			if ( canPay ) {
-				setInitializedPaymentMethod( current );
-			}
-			// update the initialized count
-			countPaymentMethodsInitializing.current--;
-			// if count remaining less than 1, then set initialized.
-			if ( countPaymentMethodsInitializing.current < 1 ) {
-				setIsInitialized( true );
-			}
-		};
-		// loop through payment methods and see what the state is
 		for ( const paymentMethodName in registeredPaymentMethods ) {
 			const current = registeredPaymentMethods[ paymentMethodName ];
-			// if in editor context then we bypass can pay check.
+
 			if ( isEditor ) {
-				updatePaymentMethod( current );
-			} else {
-				Promise.resolve(
-					current.canMakePayment( canPayArgument.current )
-				)
-					.then( ( canPay ) => {
-						if ( canPay.error ) {
-							throw new Error( canPay.error.message );
-						} else {
-							updatePaymentMethod( current, canPay );
-						}
-					} )
-					.catch( ( error ) => {
-						updatePaymentMethod( current, false );
-						if ( CURRENT_USER_IS_ADMIN ) {
-							throw new Error(
-								sprintf(
-									__(
-										// translators: %s is the error method returned by the payment method.
-										'Problem with payment method initialization: %s',
-										'woo-gutenberg-products-block'
-									),
-									error.message
-								)
-							);
-						}
-					} );
+				setInitializedPaymentMethod( current );
+				continue;
 			}
+
+			Promise.resolve( current.canMakePayment( canPayArgument.current ) )
+				.then( ( canPay ) => {
+					if ( canPay.error ) {
+						throw new Error( canPay.error.message );
+					}
+					setInitializedPaymentMethod( current, canPay );
+				} )
+				.catch( handleRegistrationError );
 		}
-	}, [ isInitialized, isEditor ] );
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ isEditor ] );
 
-	// once all payment methods have been initialized, resort to be in the same
-	// order as registered and then set via the dispatcher.
+	// As payment methods are initialized once promises resolve, see if we're finished, resort to be
+	// in the same order as registered and then set via the dispatcher.
 	// Note: registeredPaymentMethods is not a dependency because this will not
 	// change in the life of the hook, it comes from an externally set value.
 	useEffect( () => {
-		if ( isInitialized ) {
-			const reSortByRegisteredOrder = () => {
-				const newSet = {};
-				for ( const paymentMethodName in registeredPaymentMethods ) {
-					if ( initializedPaymentMethods[ paymentMethodName ] ) {
-						newSet[ paymentMethodName ] =
-							initializedPaymentMethods[ paymentMethodName ];
-					}
+		if (
+			! isInitialized &&
+			Object.keys( initializedPaymentMethods ).length ===
+				Object.keys( registeredPaymentMethods ).length
```

Comment:
This replaces the count ref that was there previously. `initializedPaymentMethods` contains the payment methods regardless of the result from the promise, so we can rely on its size to know if we're done or not.